openvdb/points/PointScatter.h

```diff
@@ -0,0 +1,676 @@
+///////////////////////////////////////////////////////////////////////////
+//
+// Copyright (c) 2012-2017 DreamWorks Animation LLC
+//
+// All rights reserved. This software is distributed under the
+// Mozilla Public License 2.0 ( http://www.mozilla.org/MPL/2.0/ )
+//
+// Redistributions of source code must retain the above copyright
+// and license notice and the following restrictions and disclaimer.
+//
+// *     Neither the name of DreamWorks Animation nor the names of
+// its contributors may be used to endorse or promote products derived
+// from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+// IN NO EVENT SHALL THE COPYRIGHT HOLDERS' AND CONTRIBUTORS' AGGREGATE
+// LIABILITY FOR ALL CLAIMS REGARDLESS OF THEIR BASIS EXCEED US$250.00.
+//
+///////////////////////////////////////////////////////////////////////////
+//
+/// @author Nick Avramoussis
+///
+/// @file PointScatter.h
+///
+/// @brief Various point scattering methods for generating VDB Points.
+///
+///  Note that this architecture is heavily based off the similar
+///  functionality for more generic position scattering in
+///  tools/PointScatter.h, with an additional set of free functions for
+///  convenience.
+///
+///  All random number calls are made to the same generator to produce
+///  temporarily consistent results in relation to the provided seed. This
+///  comes with some multi-threaded performance trade-offs.
+///
+#ifndef OPENVDB_POINTS_POINT_SCATTER_HAS_BEEN_INCLUDED
+#define OPENVDB_POINTS_POINT_SCATTER_HAS_BEEN_INCLUDED
+
+#include <algorithm>
+#include <thread>
+
+#include <openvdb/openvdb.h>
+#include <openvdb/tree/LeafManager.h>
+#include <openvdb/tools/Prune.h>
+#include <openvdb/util/NullInterrupter.h>
+
+#include "PointDataGrid.h"
+#include "AttributeArray.h"
+#include "PointCount.h"
+
+#include <tbb/parallel_sort.h>
+#include <tbb/parallel_for.h>
+
+namespace openvdb {
+OPENVDB_USE_VERSION_NAMESPACE
+namespace OPENVDB_VERSION_NAME {
+namespace points {
+
+/// @brief Uniformly scatter a total amount of points in active regions
```

Comment:
Also, I would put the scattering classes (UniformPointScatter, etc) inside a point_scatter_internal namespace and add doxygen comments around the free functions only. That way it's possible to adjust the implementation of the internal namespace classes without breaking the exposed API and leave users more accepting of breaking API changes to the internal namespace classes.

Reply:
An answer to all your comments on structure: 

The only reason I chose this inherited class structure is to match the current implementation in tools/PointScatter.h. If you're suggesting adding an internal namespace and removing the BaseScatter class I'm more than happy to change all of this into a set of external free functions (the ones that currently exist) and a set of functions with corresponding functors within an internal namespace. This will be more akin to files like PointCount/PointConvert. The SOP would then have a small struct wrapping these so it can be passed to Houdini's runtime type executer calls. There's not enough consistency to make an informed decision so it feels more like personal preference. If matching the current PointScatter.h isn't useful then I suggest I remove all the classes and move most the functionality into the main free functions.